src/lib/core/redux/slices/rtcAnalytics.ts

```diff
@@ -0,0 +1,202 @@
+        getOutput: (value) => ({ clientId: value }),
+    },
+    deviceId: {
+        actionType: "",
+        rtcEventName: "deviceId",
+        getValue: (state: RootState) => selectDeviceId(state),
+        getOutput: (value) => ({ deviceId: value }),
+    },
+    externalId: {
+        actionType: "",
+        rtcEventName: "externalId",
+        getValue: (state: RootState) => selectAppExternalId(state),
+        getOutput: (value) => ({ externalId: value }),
+    },
+    organizationId: {
+        actionType: "",
+        rtcEventName: "organizationId",
+        getValue: (state: RootState) => selectOrganizationId(state),
+        getOutput: (value) => ({ organizationId: value }),
+    },
+    signalConnectionStatus: {
+        actionType: "",
+        rtcEventName: "signalConnectionStatus",
+        getValue: (state: RootState) => selectSignalStatus(state),
+        getOutput: (value) => ({ status: value }),
+    },
+    rtcConnectionStatus: {
+        actionType: "",
+        rtcEventName: "rtcConnectionStatus",
+        getValue: (state: RootState) => selectRtcStatus(state),
+        getOutput: (value) => ({ status: value }),
+    },
+    userRole: {
+        actionType: "",
+        rtcEventName: "userRole",
+        getValue: (state: RootState) => selectLocalParticipantUserRole(state),
+        getOutput: (value) => ({ userRole: value }),
+    },
+};
+
+const makeComparable = (value: unknown) => {
+    if (typeof value === "object") return JSON.stringify(value);
+
+    return value;
+};
+
+/**
+ * Reducer
+ */
+
+export interface rtcAnalyticsState {
+    reportedValues: { [key: string]: unknown };
+}
+
+const initialState: rtcAnalyticsState = {
+    reportedValues: {},
+};
+
+export const rtcAnalyticsSlice = createSlice({
+    initialState,
+    name: "rtcAnalytics",
+    reducers: {
+        updateReportedValues(state, action: PayloadAction<{ rtcEventName: string; value: unknown }>) {
+            state.reportedValues[action.payload.rtcEventName] = action.payload.value;
+        },
+    },
+});
+
+export const doRtcAnalyticsCustomEventsInitialize = createAppThunk(() => (dispatch, getState) => {
+    const state = getState();
+    const rtcManager = selectRtcConnectionRaw(state).rtcManager;
+
+    if (!rtcManager) return;
+
+    Object.values(rtcAnalyticsCustomEvents).forEach(({ rtcEventName, getValue, getOutput }) => {
+        const value = getValue(state);
+        const output = { ...(getOutput(value) as Record<string, unknown>), _time: Date.now() };
+
+        const comparableValue = makeComparable(value);
+
+        if (state.rtcAnalytics.reportedValues?.[rtcEventName] !== comparableValue) {
+            rtcManager.sendStatsCustomEvent(rtcEventName, output);
+            dispatch(updateReportedValues({ rtcEventName, value }));
+        }
+    });
+});
+
+/**
+ * Action creators
+ */
+export const { updateReportedValues } = rtcAnalyticsSlice.actions;
+
+startAppListening({
+    predicate: (_action) => {
```

Comment:
Wonder if we need this predicate at all. Why not put everything in the effect and return early if it is an irrelevant action?

Reply:
There is also the option of using an action `matcher` as described here: https://redux-toolkit.js.org/api/createListenerMiddleware#startlistening

Reply:
I agree that we should use the action `matcher`. We also need to change the events map so that we don't use strings, as it's super easy to forget when we update/change action creator names. I've created a ticket for this and I'm planning on picking that up asap. 

So this can be merged as it is.